# Data Subset Creation
This document serves to guide people through the process of creating a subset from the Air_Quality CSV file. The subset focuses on the NO2, O3, and fine particle values in Manhattan and Brooklyn. By filtering out the rest of the data,  between the two regions can be made, alongside analysis of the amount each molecules presence over a 10 year period. The first part will discuss setup, including sourcing the data, opening colabs, mounting your Google Drive and importing the tools necessary for filtering. The next section tackles creation of the subset through the removal of columns and rows that are unecessary. Lastly, creating and exporting the new subset as a CSV file will be explained so that it can be saved and used later.
## Setup
#### Getting Data
1.    Navigate to (https://catalog.data.gov/dataset/air-quality)
2.    Under **Downloads & Resources**, download the comma seperated values file
3.    Upload this file (Air_Quality.csv) to your Google Drive
#### Colab Setup
**Every time you need to run another set of code, click the *+code* button near the top**

**If any errors happen, please retrace the steps to try and troubleshoot.**
1.    Navigate to (https://colab.google/)
2.    Click **New Notebook**
3.    In the new notebook mount your google drive via the following code:

In [1]:
from google.colab import drive
drive.mount('/content/gdrive')


Mounted at /content/gdrive


4.    Click the play button next to the code   
5.    When prompted, click **Connect to Google Drive**
6.    Sign in to your Google Account, click **Continue**
7.    Select as many or as little permissions as you want and click **Continue**
8.    After it has mounted it should say under the code: *Mounted at /content/gdrive*
9.    Navigate to your Google Drive, and move **Air_Quality.csv** into the folder titled **Colab Notebooks**
10.   In Colab, click on the folder icon on the left sidebar; if the folder **gdrive** shows up, click on it; if not, click on **content**, then click on **gdrive**
11.   Click on **My Drive**, then **Colab Notebooks**, under which you should see Air_Quality.csv
12.   Next import pandas and numpy by running the following code:

In [2]:
import numpy as np
import pandas as pd

Anytime a pandas function is used, we will use pd and np for any numpy functions

13.   Import the CSV file by running the following code:

In [3]:
pd.read_csv('gdrive/My Drive/Colab Notebooks/Air_Quality.csv')

Unique ID  Indicator ID                                         Name  \
0         172653           375                       Nitrogen dioxide (NO2)   
1         172585           375                       Nitrogen dioxide (NO2)   
2         336637           375                       Nitrogen dioxide (NO2)   
3         336622           375                       Nitrogen dioxide (NO2)   
4         172582           375                       Nitrogen dioxide (NO2)   
...          ...           ...                                          ...   
16213     130750           647            Outdoor Air Toxics - Formaldehyde   
16214     130780           647            Outdoor Air Toxics - Formaldehyde   
16215     131020           652  Cardiac and respiratory deaths due to Ozone   
16216     131026           652  Cardiac and respiratory deaths due to Ozone   
16217     325247           643                Annual vehicle miles traveled   

                            Measure Measure Info Geo Type Name  Geo Join ID  \
0                              Mean          ppb         UHF34          203   
1                              Mean          ppb         UHF34          203   
2                              Mean          ppb         UHF34          204   
3                              Mean          ppb         UHF34          103   
4                              Mean          ppb         UHF34          104   
...                             ...          ...           ...          ...   
16213  Annual average concentration        µg/m3         UHF42          211   
16214  Annual average concentration        µg/m3       Borough            5   
16215         Estimated annual rate  per 100,000         UHF42          504   
16216         Estimated annual rate  per 100,000       Borough            5   
16217                 million miles      per km2            CD          107   

                           Geo Place Name          Time Period  Start_Date  \
0      Bedford Stuyvesant - Crown Heights  Annual Average 2011  12/01/2010   
1      Bedford Stuyvesant - Crown Heights  Annual Average 2009  12/01/2008   
2                           East New York  Annual Average 2015  01/01/2015   
3                      Fordham - Bronx Pk  Annual Average 2015  01/01/2015   
4                    Pelham - Throgs Neck  Annual Average 2009  12/01/2008   
...                                   ...                  ...         ...   
16213             Williamsburg - Bushwick                 2005  01/01/2005   
16214                       Staten Island                 2005  01/01/2005   
16215           South Beach - Tottenville            2005-2007  01/01/2005   
16216                       Staten Island            2005-2007  01/01/2005   
16217               Upper West Side (CD7)                 2016  01/01/2016   

       Data Value  Message  
0           25.30      NaN  
1           26.93      NaN  
2           19.09      NaN  
3           19.76      NaN  
4           22.83      NaN  
...           ...      ...  
16213        3.10      NaN  
16214        2.30      NaN  
16215        7.50      NaN  
16216        7.80      NaN  
16217       50.00      NaN  

[16218 rows x 12 columns]

14.   A table should pop up with columns that include: Name, Geo Place Name, and Message
15.   Check to ensure the table has 16218 rows × 12 columns; if not, it is the wrong file.

You are now done with the setup portion, and we can move on.
## Subset Creation
1.    Step one is creating a data frame object that houses the CSV file. This is accomplished by using a word or phrase with the = sign and the code from above. Anything can be used here to define the data frame; the example below defines it as "AirData."

In [4]:
AirData = pd.read_csv('gdrive/My Drive/Colab Notebooks/Air_Quality.csv')

2. We can use the new data frame name until we create another data frame.
3.    Feel free to test this out by using AirData followed by a command such as *AirData.columns*, which displays the label of each column in the data frame.
4.    After this, we will eliminate unnecessary columns using the drop function. Two things should be done before executing the function: first, a new data frame name, in the example below, "Cleaning_1." Second, .copy() needs to be added to the end of the function to ensure that the original data frame remains intact if you need to use it again.
5.    In this case we want to get rid of the columns labeled: unique ID, indicator ID, geo type name, start date, message, and geo join ID. These will be removed since they are not relevant for the specific purpose of this subset.
5.    Your code for this should look similar to the one below, with the  specific columns we want to remove being identified by name.

In [5]:
Cleaning_1 = AirData.drop(columns= ['Unique ID', 'Indicator ID','Geo Type Name','Start_Date','Message', 'Geo Join ID']).copy()

7.    Next, type Cleaning_1 and run it. A table with 16218 rows x 6 Columns should appear, indicating that the drop function correctly removed the six unnecessary columns.

In [6]:
Cleaning_1

Name  \
0                           Nitrogen dioxide (NO2)   
1                           Nitrogen dioxide (NO2)   
2                           Nitrogen dioxide (NO2)   
3                           Nitrogen dioxide (NO2)   
4                           Nitrogen dioxide (NO2)   
...                                            ...   
16213            Outdoor Air Toxics - Formaldehyde   
16214            Outdoor Air Toxics - Formaldehyde   
16215  Cardiac and respiratory deaths due to Ozone   
16216  Cardiac and respiratory deaths due to Ozone   
16217                Annual vehicle miles traveled   

                            Measure Measure Info  \
0                              Mean          ppb   
1                              Mean          ppb   
2                              Mean          ppb   
3                              Mean          ppb   
4                              Mean          ppb   
...                             ...          ...   
16213  Annual average concentration        µg/m3   
16214  Annual average concentration        µg/m3   
16215         Estimated annual rate  per 100,000   
16216         Estimated annual rate  per 100,000   
16217                 million miles      per km2   

                           Geo Place Name          Time Period  Data Value  
0      Bedford Stuyvesant - Crown Heights  Annual Average 2011       25.30  
1      Bedford Stuyvesant - Crown Heights  Annual Average 2009       26.93  
2                           East New York  Annual Average 2015       19.09  
3                      Fordham - Bronx Pk  Annual Average 2015       19.76  
4                    Pelham - Throgs Neck  Annual Average 2009       22.83  
...                                   ...                  ...         ...  
16213             Williamsburg - Bushwick                 2005        3.10  
16214                       Staten Island                 2005        2.30  
16215           South Beach - Tottenville            2005-2007        7.50  
16216                       Staten Island            2005-2007        7.80  
16217               Upper West Side (CD7)                 2016       50.00  

[16218 rows x 6 columns]

8. After confirming that step one of cleaning is done, we need to remove many unnecessary rows. This will be done using the **.isin** function, which will check all rows in a specific column to see if they contain something that we specified, in this case, specific molecules on the "Name" column. Isolating O3, NO2, and fine particles allows us to use these molecules for comparisons later on without having to sort through unnecessary date like Cardiac and respiratory deaths due to Ozone that is not the focus of this subset.
9.    Before running the code, which should look like the one below, make sure to create a new data frame object and add the copy function at the end. In this case, the latest data frame object is Cleaning_2."
10.    Notice that the specific columns and objects are labeled in the following example code to specify what we want to keep.

In [7]:
Cleaning_2 = Cleaning_1[Cleaning_1['Name'].isin(['Nitrogen dioxide (NO2)', 'Fine particles (PM 2.5)', 'Ozone (O3)'])].copy()

11. Like before, run the code "Cleaning_2" and check if you have a table with 12972 rows x 6 columns.

In [8]:
Cleaning_2

Name Measure Measure Info  \
0      Nitrogen dioxide (NO2)    Mean          ppb   
1      Nitrogen dioxide (NO2)    Mean          ppb   
2      Nitrogen dioxide (NO2)    Mean          ppb   
3      Nitrogen dioxide (NO2)    Mean          ppb   
4      Nitrogen dioxide (NO2)    Mean          ppb   
...                       ...     ...          ...   
16207  Nitrogen dioxide (NO2)    Mean          ppb   
16208  Nitrogen dioxide (NO2)    Mean          ppb   
16209              Ozone (O3)    Mean          ppb   
16210  Nitrogen dioxide (NO2)    Mean          ppb   
16211  Nitrogen dioxide (NO2)    Mean          ppb   

                           Geo Place Name          Time Period  Data Value  
0      Bedford Stuyvesant - Crown Heights  Annual Average 2011       25.30  
1      Bedford Stuyvesant - Crown Heights  Annual Average 2009       26.93  
2                           East New York  Annual Average 2015       19.09  
3                      Fordham - Bronx Pk  Annual Average 2015       19.76  
4                    Pelham - Throgs Neck  Annual Average 2009       22.83  
...                                   ...                  ...         ...  
16207                       Staten Island       Winter 2008-09       21.63  
16208                       Staten Island       Winter 2017-18       18.92  
16209                         Sunset Park          Summer 2014       30.60  
16210                         Sunset Park       Winter 2008-09       30.17  
16211                         Sunset Park       Winter 2013-14       27.68  

[12972 rows x 6 columns]

12. The next code set is the same as before, with a few changes. First, the new column that is specified is Geo Place Name, and we are isolating Queens and Manhattan for our objects.
13. Add a new data frame name and the copy function at the end. Here, it is "Cleaning_3"

In [9]:
Cleaning_3 = Cleaning_2[Cleaning_2['Geo Place Name'].isin(['Manhattan', 'Queens'])].copy()

14. Once again, run Cleaning_3 to make sure that you have a 184 x 6 table with data.

In [10]:
Cleaning_3

Name Measure Measure Info Geo Place Name  \
232     Nitrogen dioxide (NO2)    Mean          ppb         Queens   
351     Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
352     Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
374     Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
375     Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
...                        ...     ...          ...            ...   
15269   Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
15270   Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
15285   Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
15378  Fine particles (PM 2.5)    Mean       mcg/m3      Manhattan   
15383  Fine particles (PM 2.5)    Mean       mcg/m3      Manhattan   

               Time Period  Data Value  
232    Annual Average 2020       14.82  
351    Annual Average 2015       25.18  
352    Annual Average 2020       20.54  
374            Summer 2009       30.87  
375         Winter 2014-15       28.69  
...                    ...         ...  
15269          Summer 2014       22.69  
15270       Winter 2013-14       32.41  
15285  Annual Average 2019       21.71  
15378  Annual Average 2019        8.31  
15383       Winter 2013-14       14.13  

[184 rows x 6 columns]

15. The next code set is the same as before, with a few changes. First, the new column that is specified is Time Period, and we are isolating all of the Annual Averages for our objects. This code is longer since you need to type out Annual Average 2009, Annual Average 2010, and so on until Annual Average 2021.  
16. Add a new data frame name and the copy function at the end. Here, it is "AirDataSubset" since it is the final data frame that we will create.

In [11]:
AirDataSubset = Cleaning_3[Cleaning_3['Time Period'].isin(['Annual Average 2009','Annual Average 2010', 'Annual Average 2011', 'Annual Average 2012', 'Annual Average 2013', 'Annual Average 2014', 'Annual Average 2015', 'Annual Average 2016', 'Annual Average 2017', 'Annual Average 2018', 'Annual Average 2019', 'Annual Average 2020', 'Annual Average 2021'])].copy()

17. Run AirDataSubset and scroll through to make sure there arent any rows with "Summer" or "Winter".

In [12]:
AirDataSubset

Name Measure Measure Info Geo Place Name  \
232     Nitrogen dioxide (NO2)    Mean          ppb         Queens   
351     Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
352     Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
664    Fine particles (PM 2.5)    Mean       mcg/m3      Manhattan   
1560   Fine particles (PM 2.5)    Mean       mcg/m3      Manhattan   
1737   Fine particles (PM 2.5)    Mean       mcg/m3         Queens   
1738   Fine particles (PM 2.5)    Mean       mcg/m3         Queens   
2039   Fine particles (PM 2.5)    Mean       mcg/m3      Manhattan   
2041   Fine particles (PM 2.5)    Mean       mcg/m3      Manhattan   
4839    Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
10332  Fine particles (PM 2.5)    Mean       mcg/m3         Queens   
12291   Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
12292   Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
12294   Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
13153   Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
13228   Nitrogen dioxide (NO2)    Mean          ppb         Queens   
13229   Nitrogen dioxide (NO2)    Mean          ppb         Queens   
13230   Nitrogen dioxide (NO2)    Mean          ppb         Queens   
13453   Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
13457  Fine particles (PM 2.5)    Mean       mcg/m3      Manhattan   
13462  Fine particles (PM 2.5)    Mean       mcg/m3      Manhattan   
13604   Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
13623   Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
13740   Nitrogen dioxide (NO2)    Mean          ppb         Queens   
13748   Nitrogen dioxide (NO2)    Mean          ppb         Queens   
13755  Fine particles (PM 2.5)    Mean       mcg/m3         Queens   
13760  Fine particles (PM 2.5)    Mean       mcg/m3         Queens   
13763   Nitrogen dioxide (NO2)    Mean          ppb         Queens   
13764   Nitrogen dioxide (NO2)    Mean          ppb         Queens   
13766   Nitrogen dioxide (NO2)    Mean          ppb         Queens   
13774  Fine particles (PM 2.5)    Mean       mcg/m3      Manhattan   
13781   Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
13797  Fine particles (PM 2.5)    Mean       mcg/m3      Manhattan   
13927  Fine particles (PM 2.5)    Mean       mcg/m3      Manhattan   
13928  Fine particles (PM 2.5)    Mean       mcg/m3      Manhattan   
13998  Fine particles (PM 2.5)    Mean       mcg/m3      Manhattan   
14175  Fine particles (PM 2.5)    Mean       mcg/m3         Queens   
14178  Fine particles (PM 2.5)    Mean       mcg/m3         Queens   
14186  Fine particles (PM 2.5)    Mean       mcg/m3         Queens   
14190  Fine particles (PM 2.5)    Mean       mcg/m3         Queens   
14198  Fine particles (PM 2.5)    Mean       mcg/m3         Queens   
14853   Nitrogen dioxide (NO2)    Mean          ppb         Queens   
15057  Fine particles (PM 2.5)    Mean       mcg/m3         Queens   
15058  Fine particles (PM 2.5)    Mean       mcg/m3         Queens   
15059   Nitrogen dioxide (NO2)    Mean          ppb         Queens   
15060   Nitrogen dioxide (NO2)    Mean          ppb         Queens   
15063   Nitrogen dioxide (NO2)    Mean          ppb         Queens   
15075  Fine particles (PM 2.5)    Mean       mcg/m3         Queens   
15201  Fine particles (PM 2.5)    Mean       mcg/m3      Manhattan   
15268   Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
15285   Nitrogen dioxide (NO2)    Mean          ppb      Manhattan   
15378  Fine particles (PM 2.5)    Mean       mcg/m3      Manhattan   

               Time Period  Data Value  
232    Annual Average 2020       14.82  
351    Annual Average 2015       25.18  
352    Annual Average 2020       20.54  
664    Annual Average 2014       10.85  
1560   Annual Average 2013       10.73  
1737   Annual Average 2009       10.05  
1738   Annual Average 2020  

Now that we have isolated NO3, O3, and fine particles in Queens and Manhattan, we can move on to exporting this new subset as a CSV file.
## Exporting the Subset
1.    Exporting is rather simple using only the following bit of code, with index = false stopping pandas from adding indices to the file.
2.    Your code should look similar to the one below:

In [13]:
AirDataSubset.to_csv("AirDataSubset.csv", index=False)

2.    After this is run, click on the folder icon and click on the folder icon on the left-hand side. The new CSV should be there, if not, follow the steps from the Colab set up portion to find it in the Colab Notebook part of your Google Drive. from here you can download it to your computer.
3.    You now have the subset and can use it to compare different bits of the data it provides.